### ISSUE TITLE
Stack overflow during grading occurs sooner in the browser than in batch mode

### ISSUE THREAD

**Issue by fpottier** [ocaml-sf/learn-ocaml#242]
The title says it all: I have written a grader which works in batch mode (when executed via `learn-ocaml grade`) but fails in the browser (Firefox) with a `Stack_overflow` exception. This happens when grading the solution of my exercise `nondet_monad_cont`, which currently is sitting in a pull request in `learn-ocaml-corpus`. The second half of Question 4 (laziness) is where the grader fails.

Is it possible that the compilation to JavaScript does not do tail call optimization?

Is it possible to somehow ensure that the stack limit is the same under both environments (batch mode and in-browser mode)?

Is it possible for each exercise to indicate what stack size it needs?

**Comment by ejgallego** [comment]
> Is it possible that the compilation to JavaScript does not do tail call optimization?

Indeed tail call is limited in js_of_ocaml; see the docs.

We use a workaround for Coq https://github.com/ejgallego/jscoq/blob/v8.10/etc/patches/trampoline.patch

**Comment by fpottier** [comment]
Thanks Emilio.

Interesting workaround, which I don't fully understand: doesn't the fact that `trampoline` itself is tail-recursive still create a problem?

Besides, I suppose that this workaound is applicable to code that is already written in monadic style, but this does not really solve the problem with "normal" code.

**Comment by ejgallego** [comment]
> Oh, I see, this pattern is recognized by js_of_ocaml.

Indeed, what the workaround does is to help JSOO realize about the tail-rec; for teaching this is a problem indeed.

**Comment by fpottier** [comment]
Does anyone have a suggestion to fix this problem? It's now exposed on the public site.

**Comment by erikmd** [comment]
Hi @fpottier, that's an annoying problem indeed.

FWIW, I recently identified another `Stack_overflow` problem after I merged learn-ocaml master in the **learn-ocaml-editor** branch (deployed at https://pfitaxel.github.io/pfitaxel-demo/):
* the source of the issue was painful to identify (as there is no backtrace or so) but I pushed a workaround in https://github.com/pfitaxel/learn-ocaml/commit/d337cccf2b276ba51c37784d4d9f69de7f0321bd;
* still, this won't be helpful for you as it is orthogonal to your issue: to sum up, mine was directly trigerred by [a call to the function `descrs_from_string` on a large string…](https://github.com/pfitaxel/learn-ocaml/blob/d337cccf2b276ba51c37784d4d9f69de7f0321bd/src/repo/learnocaml_exercise.ml#L335-L342)
* I didn't bisect manually but the first commit that certainly introduced that issue (which didn't occur with the first versions of learn-ocaml/learn-ocaml-editor) is https://github.com/ocaml-sf/learn-ocaml/commit/163b0d12ce6032ceddb40f80a898903fb9d31f49
* And unfortunately I don't have a good solution to fix it for the time being, the commit `d337ccc` being just a raw workaround (replacing the large string with an error message) :-/

**Comment by hferee** [comment]
I haven't had any stack overflow issue since I'm using v1.0.0 and I have tried several stress tests, similar to those mentioned above, and no overflow.

I'm not sure what changed this (a change in jsofocaml maybe?), but I reckon that this issue can be closed, as well as #524.

**Comment by yurug** [comment]
Let's close it then! Thanks for the report @hferee
